### PR TITLE
Add debugg logging for DataProtectorTokenProvider.ValidationAsync

### PR DESCRIPTION
**Pull request by dpr-dev**
changes [#5826](https://github.com/aspnet/AspNetCore/issues/5826)

### CODE REVIEW

File: src/Identity/Core/src/DataProtectorTokenProvider.cs [hunk 1]
```diff
@@ -110,38 +125,58 @@ public virtual async Task<bool> ValidateAsync(string purpose, string token, User
...
                     }
-                    return stamp == "";
+
+                    return stampIsEmpty; 
                 }
             }
-            // ReSharper disable once EmptyGeneralCatchClause
-            catch
-            {
-                // Do not leak exception
```
Comment (HaoK) [inline src/Identity/Core/src/DataProtectorTokenProvider.cs]: @blowdart how important is it that we don't leak exceptions here?  Given the comment do we wante to err on the side of not logging the exact exception message?

Comment (blowdart) [reply src/Identity/Core/src/DataProtectorTokenProvider.cs]: Yes, this needs to return. We cannot return the exception message, only that an exception occured.

File: src/Identity/Core/src/DataProtectorTokenProvider.cs [hunk 2]
```diff
@@ -110,38 +125,58 @@ public virtual async Task<bool> ValidateAsync(string purpose, string token, User
                     var expirationTime = creationTime + Options.TokenLifespan;
                     if (expirationTime < DateTimeOffset.UtcNow)
                     {
+                        Logger.LogWarning(0, "ValidateAsync failed: the expiration time is invalid.");
```
Comment (HaoK) [inline src/Identity/Core/src/DataProtectorTokenProvider.cs]: I know its a pain, but the pattern we are supposed to use for logging going forward is structured logging, see: https://github.com/aspnet/AspNetCore/blob/master/src/Security/Authentication/Core/src/LoggingExtensions.cs

Can you put these new log messages into a new LoggingExtensions class like that?

Comment (dpr-dev) [reply src/Identity/Core/src/DataProtectorTokenProvider.cs]: Yes of course. thank. I did not know about it

File: src/Identity/Core/src/DataProtectorTokenProvider.cs (hunk 1 above)
Comment (dpr-dev) [reply src/Identity/Core/src/DataProtectorTokenProvider.cs]: > Yes, this needs to return. We cannot return the exception message, only that an exception occured.

okay. Should we write a debug message when an exception occurs ? Or not ?

Comment (HaoK) [reply src/Identity/Core/src/DataProtectorTokenProvider.cs]: No I think we should just log that an exception has occurred without any of the details in the exception, basically go back to the general catch

**Comment by dpr-dev** [comment]
@HaoK, please, review new commits

File: src/Identity/Core/src/LoggingExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            _invalidExpirationTime = LoggerMessage.Define(
+                eventId: new EventId(0, "InvalidExpirationTime"),
+                logLevel: LogLevel.Information,
+                formatString: "ValidateAsync failed: the expiration time is invalid.");
+
+            _userIdsNotEquals = LoggerMessage.Define(
+                eventId: new EventId(1, "UserIdsNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual UserId aren't equal");
```
Comment (HaoK) [inline src/Identity/Core/src/LoggingExtensions.cs]: nit: "ValidateAsync failed: did not find expected UserId."  Note there's also a period missing from this one

File: src/Identity/Core/src/LoggingExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+            _purposeNotEquals = LoggerMessage.Define(
+                eventId: new EventId(2, "PurposeNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual purpose aren't equal.");
```
Comment (HaoK) [inline src/Identity/Core/src/LoggingExtensions.cs]: Nit: "ValidateAsync failed: did not find expected purpose."

File: src/Identity/Core/src/LoggingExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+            _unexpectedEndOfInput = LoggerMessage.Define(
+                eventId: new EventId(3, "UnexpectedEndOfInput"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: unexpected end of input.");
+
+            _securityStampNotEquals = LoggerMessage.Define(
+                eventId: new EventId(4, "SecurityStampNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual stamp aren't equal.");
```
Comment (HaoK) [inline src/Identity/Core/src/LoggingExtensions.cs]: Nit: "ValidateAsync failed: did not find expected security stamp."

**Comment by HaoK** [comment]
@blowdart can we log additional info for some of the tokens?  i.e. expected/actual security stamps/user ids/purposes, those would help a lot with debugging...  purpose doesn't really have any PII, security stamp is just a random GUID, user id might be more of an issue.

**Comment by blowdart** [comment]
Purpose isn't a secret so sure, you can log that. The rest is PII.

File: src/Identity/Core/src/DataProtectorTokenProvider.cs [hunk 6]
```diff
@@ -110,38 +126,60 @@ public virtual async Task<bool> ValidateAsync(string purpose, string token, User
...
                     if (userId != actualUserId)
                     {
+                        Logger.UserIdsNotEquals();
                         return false;
                     }
+
                     var purp = reader.ReadString();
                     if (!string.Equals(purp, purpose))
                     {
+                        Logger.PurposeNotEquals();
```
Comment (HaoK) [inline src/Identity/Core/src/DataProtectorTokenProvider.cs]: Can you pass in purp and purpose to the log and print that out in the log, that should help diagnose when the wrong token is used

File: src/Identity/Core/src/LoggingExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            _invalidExpirationTime(logger, null);
+        }
+
+        public static void UserIdsNotEquals(this ILogger logger)
+        {
+            _userIdsNotEquals(logger, null);
+        }
+
+        public static void PurposeNotEquals(this ILogger logger)
```
Comment (HaoK) [inline src/Identity/Core/src/LoggingExtensions.cs]: Take in expectedPurpose and actualPurpose and display those to help debugging

Comment (dpr-dev) [reply src/Identity/Core/src/LoggingExtensions.cs]: Ok, I will do it. Thank you for your suggestions

**Comment by HaoK** [comment]
Almost there looks like you just need to update the ref assembly to account for the new logger property, see https://dev.azure.com/dnceng/public/_build/results?buildId=227250&view=logs&jobId=2fe95fc5-ae3d-58fd-c2b8-46bb81541875

**Comment by HaoK** [comment]
Thanks @dpr-dev !

**Pull request merged**